Event log cannot be read.

```
2017-03-28 13:44:38.182406 UTC [ERROR] [W0rPv7IF3zTPLiEHH8Lv1b0e] Action EventLog ActionPersistExcept (PersistError404 {persistErrorMessage = "Idea"})
```

First guess: moving ideas around changes their keys, so the ideas in their old places from the log files cannot be found any more.  (Needs confirmation.)

--- a/Makefile
+++ b/Makefile
@@ -110,7 +110,7 @@ testxlsxclass:
 	rm -f cookie-jar
 
 tags: .phony
-	hasktags -b src/ tests/ exec/ dist/build/autogen/
+	hasktags -b src/ tests/ exec/
 
 grepi.%:
 	git grep -Hni $* src tests exec

--- a/aula.cabal
+++ b/aula.cabal
@@ -750,6 +750,7 @@ test-suite spec
     , wreq
     , webdriver
   other-modules:
+      ActionSpec
       AllModulesSpec
       AulaTests
       AulaTests.Stories.DSL

--- a/src/Action.hs
+++ b/src/Action.hs
@@ -144,20 +144,12 @@ module Action
 where
 
 import Codec.Picture (DynamicImage)
-import Control.Exception (SomeException, assert)
+import Control.Exception (assert)
 import Control.Lens
-import Control.Monad ((>=>), filterM, void, when)
 import Control.Monad.Reader (runReader, runReaderT)
-import Control.Monad.Except (MonadError)
 import Control.Monad.Trans.Except (runExcept)
-import Data.Char (ord)
 import Data.List as List (find)
-import Data.Maybe (isJust)
 import Data.Monoid
-import Data.String.Conversions (ST, LBS, cs)
-import Data.String (fromString)
-import Data.Typeable (Typeable)
-import Data.Foldable (forM_)
 import Prelude hiding (log)
 import Servant
 import Servant.Missing
@@ -172,6 +164,7 @@ import qualified Data.Vector as V
 
 import Action.Smtp
 import AulaMetrics
+import AulaPrelude
 import Config
     ( Config
     , GetConfig(..)
@@ -1149,19 +1142,22 @@ instance ActionM m => WarmUp m ContentCold ContentWarm where
 
 -- | for internal use only.
 class WarmUp' m a where
-    warmUp' :: KeyOf a -> m a
+    warmUp'' :: KeyOf a -> m (Maybe a)
+
+warmUp' :: (Functor m, WarmUp' m a, Show (KeyOf a)) => KeyOf a -> m (Perhaps a)
+warmUp' k = maybe (PerhapsNo (show k)) PerhapsYes <$> warmUp'' k
 
 instance ActionM m => WarmUp' m User where
-    warmUp' k = mquery (findUser k)
+    warmUp'' k = query (findUser k)
 
 instance ActionM m => WarmUp' m Topic where
-    warmUp' k = mquery (findTopic k)
+    warmUp'' k = query (findTopic k)
 
 instance ActionM m => WarmUp' m Idea where
-    warmUp' k = mquery (findIdea k)
+    warmUp'' k = query (findIdea k)
 
 instance ActionM m => WarmUp' m Comment where
-    warmUp' k = mquery (findComment k)
+    warmUp'' k = query (findComment k)
 
 currentUserCapCtx :: (ActionPersist m, ActionUserHandler m, ActionLog m) => m CapCtx
 currentUserCapCtx = userOnlyCapCtx <$> currentUser

--- a/src/Arbitrary.hs
+++ b/src/Arbitrary.hs
@@ -1223,6 +1223,9 @@ fishAvatars = unsafePerformIO fishAvatarsIO
 
 -- * event log
 
+instance (Arbitrary a) => Arbitrary (Perhaps a) where
+  arbitrary = PerhapsYes <$> arbitrary
+
 instance Arbitrary EventLog where
     arbitrary = EventLog <$> arb <*> arbWord <*> nonEmpty
       where

--- a/src/Frontend/Constant.hs
+++ b/src/Frontend/Constant.hs
@@ -7,7 +7,7 @@ import Data.String.Conversions (ST)
 -- FIXME: Search for global constants
 
 topicDescMaxLength :: Int
-topicDescMaxLength = 120
+topicDescMaxLength = 800
 
 minElabPeriod :: Int
 minElabPeriod = 1

--- a/src/Logger/EventLog.hs
+++ b/src/Logger/EventLog.hs
@@ -16,6 +16,7 @@
 module Logger.EventLog
 where
 
+import Control.Exception (assert)
 import Control.Lens
 import Data.Maybe (fromMaybe)
 import Data.String (IsString)
@@ -42,7 +43,7 @@ data EventLog = EventLog
     , _eventLogURL       :: URL
     , _eventLogItems     :: [EventLogItemWarm]
     }
-  deriving (Generic)
+  deriving (Show, Generic)
 
 -- | This type is migration-critial: we may need to support parting old values if we change it in
 -- production.  See 'CSV.ToRecord' instance(s) below.
@@ -70,24 +71,32 @@ data EventLogItemValue user topic idea comment =
 
 
 type EventLogItemCold = EventLogItem (AUID User) (AUID Topic) (AUID Idea) CommentKey
-type EventLogItemWarm = EventLogItem User Topic Idea Comment
+type EventLogItemWarm = EventLogItem (Perhaps User) (Perhaps Topic) (Perhaps Idea) (Perhaps Comment)
 
 type EventLogItemValueCold = EventLogItemValue (AUID User) (AUID Topic) (AUID Idea) CommentKey
-type EventLogItemValueWarm = EventLogItemValue User Topic Idea Comment
+type EventLogItemValueWarm = EventLogItemValue (Perhaps User) (Perhaps Topic) (Perhaps Idea) (Perhaps Comment)
 
 type ContentCold = Either3 (AUID Topic) (AUID Idea) CommentKey
-type ContentWarm = Either3 Topic Idea Comment
+type ContentWarm = Either3 (Perhaps Topic) (Perhaps Idea) (Perhaps Comment)
 
+data Perhaps a = PerhapsYes a | PerhapsNo String  -- should be @('KeyOf' a)@, but 'KeyOf' is a type
+  deriving (Show, Generic)                        -- family and cannot give us Generic, which we
+                                                  -- need at least for the 'Arbitrary' instances.
+                                                  -- (at least i *think* that was the problem.
+                                                  -- could be worth trying that again.)
 
 instance SOP.Generic EventLog
 instance SOP.Generic (EventLogItem u t i c)
 instance SOP.Generic (EventLogItemValue u t i c)
+instance SOP.Generic (Perhaps a)
 
 instance Aeson.ToJSON EventLogItemCold           where toJSON = Aeson.gtoJson
 instance Aeson.ToJSON EventLogItemValueCold      where toJSON = Aeson.gtoJson
+instance Aeson.ToJSON a => Aeson.ToJSON (Perhaps a) where toJSON = Aeson.gtoJson
 
 instance Aeson.FromJSON EventLogItemCold           where parseJSON = Aeson.gparseJson
 instance Aeson.FromJSON EventLogItemValueCold      where parseJSON = Aeson.gparseJson
+instance Aeson.FromJSON a => Aeson.FromJSON (Perhaps a) where parseJSON = Aeson.gparseJson
 
 
 -- * delivering the event log
@@ -117,38 +126,55 @@ instance CSV.ToRecord (WithURL EventLogItemWarm) where
     toRecord (WithURL domainUrl (EventLogItem ispace timestamp user ev)) = CSV.toRecord
         [ ideaSpaceCode ispace
         , showTimestamp timestamp
-        , user ^. userLogin . unUserLogin . csi
+        , showUser user
         ] <> f ev
       where
         objDesc :: ContentWarm -> ST
-        objDesc (Left3   t) = "Thema " <> t ^. topicTitle . showed . csi
-        objDesc (Middle3 i) = "Idee "  <> i ^. ideaTitle  . showed . csi
-        objDesc (Right3  c) =
-            chop $ "Verbesserungsvorschlag " <> (c ^. commentText . showed . to (take 30) . csi)
+        objDesc (Left3   (PerhapsYes t)) = "Thema " <> t ^. topicTitle . showed . csi
+        objDesc (Middle3 (PerhapsYes i)) = "Idee "  <> i ^. ideaTitle  . showed . csi
+        objDesc (Right3  (PerhapsYes c)) = chop 60 $ "Verbesserungsvorschlag " <> (c ^. commentText . showed . csi . to (chop 30))
+        objDesc (Left3   (PerhapsNo t))  = "Thema (GELÖSCHT) " <> cs t
+        objDesc (Middle3 (PerhapsNo i))  = "Idee (GELÖSCHT) " <> cs i
+        objDesc (Right3  (PerhapsNo c))  = chop 60 $ "Verbesserungsvorschlag (GELÖSCHT) " <> cs c
 
-        chop :: ST -> ST
-        chop s = if ST.length s <= 60 then s else ST.take 57 s <> "..."
+        chop :: Int -> ST -> ST
+        chop i s = assert (i >= 3) $ if ST.length s <= i then s else ST.take (i - 3) s <> "..."
+
+        objLink' :: U.Main 'AllowGetPost -> ST
+        objLink' = (domainUrl <>) . absoluteUriPath . relPath
 
         objLink :: ContentWarm -> ST
-        objLink = (domainUrl <>) . absoluteUriPath . relPath . objLink'
+        objLink (Left3   (PerhapsYes t)) = objLink' $ U.listIdeasInTopic t ListIdeasInTopicTabAll Nothing
+        objLink (Middle3 (PerhapsYes i)) = objLink' $ U.viewIdea i
+        objLink (Right3  (PerhapsYes c)) = objLink' $ U.viewIdeaAtComment' (c ^. _Key)
+        -- always add the description if there is a link!
+        objLink (Left3   (PerhapsNo _))  = nil
+        objLink (Middle3 (PerhapsNo _))  = nil
+        objLink (Right3  (PerhapsNo _))  = nil
 
-        objLink' :: ContentWarm -> U.Main 'AllowGetPost
-        objLink' (Left3   t) = U.listIdeasInTopic t ListIdeasInTopicTabAll Nothing
-        objLink' (Middle3 i) = U.viewIdea i
-        objLink' (Right3  c) = U.viewIdeaAtComment' (c ^. _Key)
+        showUser :: Perhaps User -> String
+        showUser (PerhapsYes u) = u ^. userLogin . unUserLogin . csi
+        showUser (PerhapsNo u)  = "(GELÖSCHT: " <> cs u <> ")"
 
+        showTopic :: Maybe (Perhaps Topic) -> ST
+        showTopic Nothing = "wilde Ideen"
+        showTopic (Just (PerhapsYes t)) = t ^. topicTitle . showed . csi
+        showTopic (Just (PerhapsNo  t)) = "(GELÖSCHT: " <> cs t <> ")"
+
+        f :: EventLogItemValueWarm -> CSV.Record
         f (EventLogUserCreates obj) = CSV.toRecord
             [ "legt " <> objDesc obj <> " an.", objLink obj ]
 
         f (EventLogUserEdits obj) = CSV.toRecord
             [ "bearbeitet " <> objDesc obj <> ".", objLink obj ]
 
-        f (EventLogUserMarksIdeaFeasible (Middle3 -> idea) mIsFeasible) = case mIsFeasible of
-            Nothing           -> CSV.toRecord [ "löscht Durchführbarkeitsbewertung", objLink idea ]
+        f (EventLogUserMarksIdeaFeasible (Middle3 -> idea) mIsFeasible) = CSV.toRecord $ (case mIsFeasible of
+            Nothing           -> [ "löscht Durchführbarkeitsbewertung" ]
             (Just isFeasible) -> let what = case isFeasible of
                                               IdeaFeasible    -> "durchführbar."
                                               IdeaNotFeasible -> "nicht durchführbar."
-                                 in CSV.toRecord [ "bewertet Idee als " <> what, objLink idea ]
+                                 in [ "bewertet Idee als " <> what ])
+            <> [objDesc idea, objLink idea]
 
         f (EventLogUserVotesOnIdea (Middle3 -> idea) Nothing) = CSV.toRecord
             [ "zieht seine Stimme für oder gegen " <> objDesc idea <> " zurück.", objLink idea ]
@@ -168,13 +194,13 @@ instance CSV.ToRecord (WithURL EventLogItemWarm) where
             what = objDesc (Right3 $ fromMaybe comment mcomment)
 
         f (EventLogUserDelegates ctxDesc delegate) = CSV.toRecord
-            [ "beauftragt in " <> show ctxDesc <> " an " <> delegate ^. userLogin . _UserLogin . csi
+            [ "beauftragt in " <> show ctxDesc <> " an " <> showUser delegate
             , "(kein Link verfügbar)"
             -- FIXME: there should be a link, and 'show ctxDesc' needs to be polished.
             ]
 
         f (EventLogUserWithdrawsDelegation ctxDesc delegate) = CSV.toRecord
-            [ "zieht Beauftragung in " <> show ctxDesc <> " an " <> delegate ^. userLogin . _UserLogin . csi <> " zurück"
+            [ "zieht Beauftragung in " <> show ctxDesc <> " an " <> showUser delegate <> " zurück"
             , "(kein Link verfügbar)"
             -- FIXME: there should be a link, and 'show ctxDesc' needs to be polished.
             ]
@@ -186,12 +212,9 @@ instance CSV.ToRecord (WithURL EventLogItemWarm) where
             ]
 
         f (EventLogIdeaNewLocation (Middle3 -> idea) mt1 mt2) = CSV.toRecord
-            [ "verschiebt " <> objDesc idea <> " von " <> show_ mt1 <> " nach " <> show_ mt2 <> "."
+            [ "verschiebt " <> objDesc idea <> " von " <> showTopic mt1 <> " nach " <> showTopic mt2 <> "."
             , objLink idea
             ]
-          where
-            show_ :: Maybe Topic -> ST
-            show_ mt = maybe "wilde Ideen" (view topicTitle) mt ^. showed . csi
 
         f (EventLogIdeaReachesQuorum (Middle3 -> idea)) = CSV.toRecord
             [ objDesc idea <> " erreicht das Quorum.", objLink idea ]

--- a/src/Logger/EventLog.hs
+++ b/src/Logger/EventLog.hs
@@ -51,21 +51,21 @@ data EventLogItem user topic idea comment =
   deriving (Eq, Show, Generic)
 
 data EventLogItemValue user topic idea comment =
-    EventLogUserCreates           (Either3 topic idea comment)
-  | EventLogUserEdits             (Either3 topic idea comment)
-  | EventLogUserMarksIdeaFeasible idea (Maybe IdeaJuryResultType)
-  | EventLogUserVotesOnIdea       idea (Maybe IdeaVoteValue)
-  | EventLogUserVotesOnComment    idea comment (Maybe comment) UpDown
+    EventLogUserCreates              (Either3 topic idea comment)
+  | EventLogUserEdits                (Either3 topic idea comment)
+  | EventLogUserMarksIdeaFeasible    idea (Maybe IdeaJuryResultType)
+  | EventLogUserVotesOnIdea          idea (Maybe IdeaVoteValue)
+  | EventLogUserVotesOnComment       idea comment (Maybe comment) UpDown
       -- FIXME: this should just be a comment key resp. comment, but following the type errors
       -- reveals some things that are not trivial to refactor.  the current situation is not very
       -- nice: the first comment is either the target (if a top-level comment) or the parent of the
       -- target; the second is either absent (for top-level comments) or the target.  this could be
       -- easier.
-  | EventLogUserDelegates         DScope user
-  | EventLogUserWithdrawsDelegation DScope user
-  | EventLogTopicNewPhase         topic Phase Phase
+  | EventLogUserDelegates            DScope user
+  | EventLogUserWithdrawsDelegation  DScope user
+  | EventLogTopicNewPhase            topic Phase Phase
   | EventLogIdeaNewLocation          idea (Maybe topic) (Maybe topic)
-  | EventLogIdeaReachesQuorum     idea
+  | EventLogIdeaReachesQuorum        idea
   deriving (Eq, Show, Generic)
 
 

--- a/tests/ActionSpec.hs
+++ b/tests/ActionSpec.hs
@@ -1,6 +1,5 @@
 {-# LANGUAGE OverloadedStrings    #-}
 {-# LANGUAGE ScopedTypeVariables  #-}
-{-# LANGUAGE ViewPatterns         #-}
 
 {-# OPTIONS_GHC -Wall -Werror #-}
 
@@ -23,6 +22,7 @@ spec :: Spec
 spec = do
   describe "event log" $ do
     it "works" $ do
+      pendingWith "needs acid-state snapshot and log file to run."
       l :: EventLog <- runAction readEventLog
       (ST.length . cs . show $ l) `shouldNotBe` 0
 

--- a/tests/ActionSpec.hs
+++ b/tests/ActionSpec.hs
@@ -1,0 +1,51 @@
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE ViewPatterns         #-}
+
+{-# OPTIONS_GHC -Wall -Werror #-}
+
+module ActionSpec where
+
+import           Control.Lens
+import           Control.Monad.Trans.Except
+import           Data.String.Conversions
+import qualified Data.Text as ST
+import           Test.Hspec
+
+import           Action.Implementation (Action, mkRunAction)
+import           AulaTests
+import           Config
+import           Logger.EventLog
+import           Persistent (withPersist)
+
+
+spec :: Spec
+spec = do
+  describe "event log" $ do
+    it "works" $ do
+      l :: EventLog <- runAction readEventLog
+      (ST.length . cs . show $ l) `shouldNotBe` 0
+
+
+runAction :: Action a -> IO a
+runAction action = withPersist cfg $ \rp -> do
+  let run' :: Action a -> ExceptT ServantErr IO a
+      Nat run' = mkRunAction (ActionEnv rp cfg noLog Nothing)
+
+      run :: Action a -> IO a
+      run a = runExceptT (run' a) >>= throwLeft
+
+      throwLeft :: Either ServantErr a -> IO a
+      throwLeft (Left e)  = throwIO . ErrorCall . show $ e
+      throwLeft (Right v) = pure v
+
+  run action
+  where
+    cfg :: Config
+    cfg = defaultConfig
+            & persist . persistenceImpl       .~ AcidStateOnDisk
+            & persist . dbPath                .~ "./issue-1030/state/AulaData/"
+            & logging . eventLogPath          .~ "./issue-1030/aula-events.json"
+
+    noLog :: SendLogMsg
+    noLog = SendLogMsg $ \_ -> pure ()

